File: ios/RCTWebRTC/ScreenCapturer.m
#if TARGET_OS_IOS

#include <mach/mach_time.h>

#import <ReplayKit/ReplayKit.h>
#import <WebRTC/RTCCVPixelBuffer.h>
#import <WebRTC/RTCVideoFrameBuffer.h>

#import "ScreenCapturer.h"
#import "SocketConnection.h"

const NSUInteger kMaxReadLength = 10 * 1024;

@interface Message : NSObject

@property(nonatomic, assign, readonly) CVImageBufferRef imageBuffer;
@property(nonatomic, copy, nullable) void (^didComplete)(BOOL succes, Message *message);

- (NSInteger)appendBytes:(UInt8 *)buffer length:(NSUInteger)length;

@end

@interface Message ()

@property(nonatomic, assign) CVImageBufferRef imageBuffer;
@property(nonatomic, assign) int imageOrientation;
@property(nonatomic, assign) CFHTTPMessageRef framedMessage;

@end

@implementation Message

- (instancetype)init {
    self = [super init];
    if (self) {
        self.imageBuffer = NULL;
    }

    return self;
}

- (void)dealloc {
    CVPixelBufferRelease(_imageBuffer);
}

/** Returns the amount of missing bytes to complete the message, or -1 when not enough bytes were provided to compute
 * the message length */
- (NSInteger)appendBytes:(UInt8 *)buffer length:(NSUInteger)length {
    if (!_framedMessage) {
        _framedMessage = CFHTTPMessageCreateEmpty(kCFAllocatorDefault, false);
    }

    CFHTTPMessageAppendBytes(_framedMessage, buffer, length);
    if (!CFHTTPMessageIsHeaderComplete(_framedMessage)) {
        return -1;
    }

    NSInteger contentLength =
        [CFBridgingRelease(CFHTTPMessageCopyHeaderFieldValue(_framedMessage, (__bridge CFStringRef) @"Content-Length"))
            integerValue];
    NSInteger bodyLength = (NSInteger)[CFBridgingRelease(CFHTTPMessageCopyBody(_framedMessage)) length];

    NSInteger missingBytesCount = contentLength - bodyLength;
    if (missingBytesCount == 0) {
        BOOL success = [self unwrapMessage:self.framedMessage];
        self.didComplete(success, self);

        CFRelease(self.framedMessage);
        self.framedMessage = NULL;
    }

    return missingBytesCount;
}

// MARK: Private Methods

- (CIContext *)imageContext {
    // Initializing a CIContext object is costly, so we use a singleton instead
    static CIContext *imageContext = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        imageContext = [[CIContext alloc] initWithOptions:nil];
    });

    return imageContext;
}

- (BOOL)unwrapMessage:(CFHTTPMessageRef)framedMessage {
    size_t width =
        [CFBridgingRelease(CFHTTPMessageCopyHeaderFieldValue(_framedMessage, (__bridge CFStringRef) @"Buffer-Width"))
            integerValue];
    size_t height =
        [CFBridgingRelease(CFHTTPMessageCopyHeaderFieldValue(_framedMessage, (__bridge CFStringRef) @"Buffer-Height"))
            integerValue];
    _imageOrientation = [CFBridgingRelease(
        CFHTTPMessageCopyHeaderFieldValue(_framedMessage, (__bridge CFStringRef) @"Buffer-Orientation")) intValue];

    NSData *messageData = CFBridgingRelease(CFHTTPMessageCopyBody(_framedMessage));

    // Copy the pixel buffer
    CVReturn status =
        CVPixelBufferCreate(kCFAllocatorDefault, width, height, kCVPixelFormatType_32BGRA, NULL, &_imageBuffer);
    if (status != kCVReturnSuccess) {
        NSLog(@"CVPixelBufferCreate failed");
        return false;
    }

    [self copyImageData:messageData toPixelBuffer:&_imageBuffer];

    return true;
}

- (void)copyImageData:(NSData *)data toPixelBuffer:(CVPixelBufferRef *)pixelBuffer {
    CVPixelBufferLockBaseAddress(*pixelBuffer, 0);

    CIImage *image = [CIImage imageWithData:data];
    [self.imageContext render:image toCVPixelBuffer:*pixelBuffer];

    CVPixelBufferUnlockBaseAddress(*pixelBuffer, 0);
}

@end

// MARK: -

@interface ScreenCapturer ()<NSStreamDelegate>

@property(nonatomic, strong) SocketConnection *connection;
@property(nonatomic, strong) Message *message;

@end

@implementation ScreenCapturer {
    mach_timebase_info_data_t _timebaseInfo;
    NSInteger _readLength;
    int64_t _startTimeStampNs;
}

- (instancetype)initWithDelegate:(__weak id<RTCVideoCapturerDelegate>)delegate {
    self = [super initWithDelegate:delegate];
    if (self) {
        mach_timebase_info(&_timebaseInfo);
    }

    return self;
}

- (void)setConnection:(SocketConnection *)connection {
    if (_connection != connection) {
        [_connection close];
        _connection = connection;
    }
}

- (void)startCaptureWithConnection:(SocketConnection *)connection {
    _startTimeStampNs = -1;

    self.connection = connection;
    self.message = nil;

    [self.connection openWithStreamDelegate:self];
}

- (void)stopCapture {
    self.connection = nil;
}

// MARK: Private Methods

- (void)readBytesFromStream:(NSInputStream *)stream {
    if (!stream.hasBytesAvailable) {
        return;
    }

    if (!self.message) {
        self.message = [[Message alloc] init];
        _readLength = kMaxReadLength;

        __weak __typeof__(self) weakSelf = self;
        self.message.didComplete = ^(BOOL success, Message *message) {
            if (success) {
                [weakSelf didCaptureVideoFrame:message.imageBuffer withOrientation:message.imageOrientation];
            }

            weakSelf.message = nil;
        };
    }

    uint8_t buffer[_readLength];
    NSInteger numberOfBytesRead = [stream read:buffer maxLength:_readLength];
    if (numberOfBytesRead < 0) {
        NSLog(@"error reading bytes from stream");
        return;
    }

    _readLength = [self.message appendBytes:buffer length:numberOfBytesRead];
    if (_readLength == -1 || _readLength > kMaxReadLength) {
        _readLength = kMaxReadLength;
    }
}

- (void)didCaptureVideoFrame:(CVPixelBufferRef)pixelBuffer withOrientation:(CGImagePropertyOrientation)orientation {
    int64_t currentTime = mach_absolute_time();
    int64_t currentTimeStampNs = currentTime * _timebaseInfo.numer / _timebaseInfo.denom;

    if (_startTimeStampNs < 0) {
        _startTimeStampNs = currentTimeStampNs;
    }

    RTCCVPixelBuffer *rtcPixelBuffer = [[RTCCVPixelBuffer alloc] initWithPixelBuffer:pixelBuffer];
    int64_t frameTimeStampNs = currentTimeStampNs - _startTimeStampNs;

    RTCVideoRotation rotation;
    switch (orientation) {
        case kCGImagePropertyOrientationLeft:
            rotation = RTCVideoRotation_90;
            break;
        case kCGImagePropertyOrientationDown:
            rotation = RTCVideoRotation_180;
            break;
        case kCGImagePropertyOrientationRight:
            rotation = RTCVideoRotation_270;
            break;
        default:
            rotation = RTCVideoRotation_0;
            break;
    }

    RTCVideoFrame *videoFrame = [[RTCVideoFrame alloc] initWithBuffer:rtcPixelBuffer
                                                             rotation:rotation
                                                          timeStampNs:frameTimeStampNs];

    [self.delegate capturer:self didCaptureVideoFrame:videoFrame];
}

@end

@implementation ScreenCapturer (NSStreamDelegate)

- (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode {
    switch (eventCode) {
        case NSStreamEventOpenCompleted:
            NSLog(@"server stream open completed");
            break;
        case NSStreamEventHasBytesAvailable:
            [self readBytesFromStream:(NSInputStream *)aStream];
            break;
        case NSStreamEventEndEncountered:
            NSLog(@"server stream end encountered");
            [self stopCapture];
            [self.eventsDelegate capturerDidEnd:self];
            break;
        case NSStreamEventErrorOccurred:
            NSLog(@"server stream error encountered: %@", aStream.streamError.localizedDescription);
            break;

        default:
            break;
    }
}

@end

#endif

File: ios/RCTWebRTC/ScreenCapturer.swift
import Foundation
import CoreMedia
import WebRTC
import CoreImage
import UIKit
import ReplayKit

@objc public class ScreenCapturer: RTCVideoCapturer, StreamDelegate {
    
    private var connection: SocketConnection?
    
    private var cache: Data?
    
    private var videoDecoder: CustomVideoDecoderTranscoding
    
    @objc public weak var eventsDelegate: CapturerEventsDelegate?
    
    private var startPresentationTimeStamp: Float64?
    private var presentationTimeStamps: [Float64] = []
    private var orientations: [UInt] = []
    
    @objc override public init(delegate: RTCVideoCapturerDelegate) {
        videoDecoder = CustomVideoDecoderTranscoding()
        
        super.init(delegate: delegate)
        
        Task<Void, Never> {
            for await decodedSampleBuffer in videoDecoder.decodedSampleBuffers {

                let framePresentationTimeStamp = presentationTimeStamps.removeFirst()
                let orientation = orientations.removeFirst()
                
                if startPresentationTimeStamp == nil {
                    startPresentationTimeStamp = framePresentationTimeStamp
                }
                
                let frameTimeStampNs = Int64((framePresentationTimeStamp - startPresentationTimeStamp!) * 1_000_000_000)
                
                guard let pixelBuffer = CMSampleBufferGetImageBuffer(decodedSampleBuffer) else { return }
                let rtcPixelBuffer = RTCCVPixelBuffer(pixelBuffer: pixelBuffer)
                var rotation = RTCVideoRotation._0
                switch CGImagePropertyOrientation(rawValue: UInt32(orientation))! {
                case CGImagePropertyOrientation.left:
                    rotation = ._90
                case CGImagePropertyOrientation.down:
                    rotation = ._180
                case CGImagePropertyOrientation.right:
                    rotation = ._270
                default:
                    break
                }

                let videoFrame = RTCVideoFrame(buffer: rtcPixelBuffer, rotation: rotation, timeStampNs: frameTimeStampNs)
                
                self.delegate?.capturer(self, didCapture: videoFrame)
            }
        }
    }

    @objc public func startCapture(connection: SocketConnection) {
        print("ScreenCapturer: start capture")
        self.connection = connection
        
        connection.open(with: self)
    }

    @objc public func stopCapture() {
        print("ScreenCapturer: stop capture")
        connection = nil;
    }
    
    @objc public func stream(_ aStream: Stream, handle eventCode: Stream.Event) {
        switch eventCode {
            case .openCompleted:
                print("ScreenCapturer: server stream open completed")
                break
            case .hasBytesAvailable:
                if let inputStream = aStream as? InputStream {
                    readFromStream(inputStream)
                } else {
                    print("ScreenCapturer: failed to read from non-input stream: \(aStream)")
                }
                break
            case .endEncountered:
                print("ScreenCapturer: server stream end encountered")
                self.stopCapture()
                self.eventsDelegate?.capturerDidEnd(self)
                break
            case .errorOccurred:
                print("ScreenCapturer: server stream error encountered: \(String(describing: aStream.streamError))")
                break
            default:
                break
        }
    }
    
    private func readFromStream(_ stream: InputStream) {
        if !stream.hasBytesAvailable {
            return
        }
        
        var buffer = [UInt8](repeating: 0, count: 4096)
        
        let bytesRead = stream.read(&buffer, maxLength: buffer.count)
        
        if bytesRead <= 0 {
            return
        }
        
        if cache == nil {
            cache = Data()
        }
        
        cache!.append(contentsOf: buffer[0..<bytesRead])

        // first 4 bytes indicate the length of the message
        guard cache!.count >= MemoryLayout<UInt32>.size else { return }
        
        let length = cache!.prefix(MemoryLayout<UInt32>.size).withUnsafeBytes { $0.load(as: UInt32.self).bigEndian }
                
        // wait until the message is fully delivered
        guard cache!.count >= MemoryLayout<UInt32>.size + Int(length) else { return }
        
        let messageData = cache!.subdata(in: MemoryLayout<UInt32>.size..<(MemoryLayout<UInt32>.size + Int(length)))

        cache!.removeSubrange(0..<(MemoryLayout<UInt32>.size + Int(length)))

        processMessage(messageData)
    }
    
    private func processMessage(_ message: Data) {
        let presentationTimeStampData = message.subdata(in: 0..<MemoryLayout<Float64>.size)
        let orientationData = message.subdata(in: MemoryLayout<Float64>.size..<(MemoryLayout<Float64>.size + MemoryLayout<UInt>.size))
        let imageData = message.subdata(in: (MemoryLayout<Float64>.size + MemoryLayout<UInt>.size)..<message.count)
        
        let presentationTimeStamp = presentationTimeStampData.withUnsafeBytes {
            $0.load(as: Float64.self)
        }
        presentationTimeStamps.append(presentationTimeStamp)
        
        let orientation = orientationData.withUnsafeBytes {
            $0.load(as: UInt.self)
        }
        orientations.append(orientation)
        
        videoDecoder.decode(imageData)
    }
}

class CustomVideoDecoderTranscoding {
    private let videoDecoder: VideoDecoder
    private let videoDecoderAnnexBAdaptor: VideoDecoderAnnexBAdaptor

    init() {
        videoDecoder = VideoDecoder(config: .init(realTime: true))
        videoDecoderAnnexBAdaptor = VideoDecoderAnnexBAdaptor(
            videoDecoder: videoDecoder,
            codec: .h264
        )
    }

    func decode(_ data: Data) {
        videoDecoderAnnexBAdaptor.decode(data)
    }

    var decodedSampleBuffers: AsyncStream<CMSampleBuffer> {
        videoDecoder.decodedSampleBuffers
    }
}

class CustomVideoDecoderJPEG {
    private var imageContext = CIContext(options: nil)

    private var continuation: AsyncStream<CMSampleBuffer>.Continuation?
    public let decodedSampleBuffers: AsyncStream<CMSampleBuffer>

    init() {
        (decodedSampleBuffers, continuation) = AsyncStream.makeStream(of: CMSampleBuffer.self)
    }

    func decode(_ data: Data) {
        guard let image = CIImage(data: data) else {
            print("image is nil")
            return
        }
        
        var pixelBuffer: CVPixelBuffer?

        var status = CVPixelBufferCreate(kCFAllocatorDefault, Int(image.extent.width), Int(image.extent.height), kCVPixelFormatType_32BGRA, nil, &pixelBuffer)

        if status != kCVReturnSuccess {
            print("failed to create pixel buffer")
            return
        }
        
        guard let pixelBuffer = pixelBuffer else {
            print("pixel buffer is nil")
            return
        }
        
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)

        imageContext.render(image, to: pixelBuffer)

        var sampleBuffer: CMSampleBuffer?
        var videoInfo: CMVideoFormatDescription?

        // Step 1: Create format description
        status = CMVideoFormatDescriptionCreateForImageBuffer(allocator: kCFAllocatorDefault,
                                                                imageBuffer: pixelBuffer,
                                                                formatDescriptionOut: &videoInfo)
        guard status == kCMBlockBufferNoErr, let formatDescription = videoInfo else {
            print("Error creating format description: \(status)")
            return
        }

        // Step 2: Create timing info
        var timingInfo = CMSampleTimingInfo(duration: CMTime.invalid,
                                            presentationTimeStamp: CMTime.invalid,
                                            decodeTimeStamp: CMTime.invalid)

        // Step 3: Create CMSampleBuffer
        status = CMSampleBufferCreateReadyWithImageBuffer(allocator: kCFAllocatorDefault,
                                                                imageBuffer: pixelBuffer,
                                                                formatDescription: formatDescription,
                                                                sampleTiming: &timingInfo,
                                                                sampleBufferOut: &sampleBuffer)
        if status != noErr {
            print("Error creating sample buffer: \(status)")
            return
        }

        CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly)
        
        guard let sampleBuffer = sampleBuffer else {
            print("sampleBuffer is nil")
            return
        }

        continuation?.yield(sampleBuffer)
    }
}
